src/internals/ir/callGraph.ts

```diff
@@ -0,0 +1,227 @@
+import { unreachable } from "../util";
+import { TactASTStore } from "./astStore";
+import { IdxGenerator } from "./indices";
+import {
+  AstFunctionDef,
+  AstReceiver,
+  AstContractInit,
+  AstStatement,
+  AstStaticCall,
+  AstMethodCall,
+  AstExpression,
+  AstStructFieldInitializer,
+} from "@tact-lang/compiler/dist/grammar/ast";
+
+export class CGEdge {
+  public idx: number;
+  constructor(
+    public src: number,
+    public dst: number,
+  ) {
+    this.idx = IdxGenerator.next("cg_edge");
+  }
+}
+
+export class CGNode {
+  public idx: number;
+  public inEdges: Set<number> = new Set();
+  public outEdges: Set<number> = new Set();
+  constructor(
+    public astId: number,
+    public name: string,
+  ) {
+    this.idx = IdxGenerator.next("cg_node");
+  }
+}
+
+export class CallGraph {
+  private nodeMap: Map<number, CGNode> = new Map();
+  private edgesMap: Map<number, CGEdge> = new Map();
+  build(astStore: TactASTStore): CallGraph {
+    this.addFunctionsToNodes(astStore);
+    this.analyzeFunctionCalls(astStore);
+    return this;
+  }
+
+  private addFunctionsToNodes(astStore: TactASTStore) {
+    for (const func of astStore.getFunctions()) {
+      const funcName = this.getFunctionName(func);
+      if (funcName) {
+        const node = new CGNode(func.id, funcName);
+        this.nodeMap.set(func.id, node);
+      }
+    }
+  }
+
+  private analyzeFunctionCalls(astStore: TactASTStore) {
+    for (const func of astStore.getFunctions()) {
+      const funcName = this.getFunctionName(func);
+      if (funcName) {
+        this.processStatements(func.statements, func.id);
+      }
+    }
+  }
+
+  private getFunctionName(
+    func: AstFunctionDef | AstReceiver | AstContractInit,
+  ): string | undefined {
+    switch (func.kind) {
+      case "function_def":
+        return func.name.text;
+      case "contract_init":
+        return `contract_init_${func.id}`;
+      case "receiver":
+        return `receiver_${func.id}`;
+      default:
+        unreachable(func);
+    }
+  }
+
+  private processStatements(statements: AstStatement[], callerId: number) {
```

Comment:
It is still not clear why don't we use `foldStatements` or `forEachStatement` here.

Reply:
Maybe a docstring comment explaining this would help.